tests/core/scenario/test_scenario_manager.py

```diff
@@ -813,6 +813,37 @@ def test_get_set_primary_scenario():
     assert _ScenarioManager._get_primary(cycle_1) == scenario_2
 
 
+def test_get_primary_scenarios_sorted():
+    cycle_1 = _CycleManager._create(Frequency.DAILY, name="foo1")
+    cycle_2 = _CycleManager._create(Frequency.DAILY, name="foo2")
+
+    scenario_1_cfg = Config.configure_scenario(name="C_scenario", id="scenario_3", tags=["banana", "kiwi"])
+    scenario_2_cfg = Config.configure_scenario(name="B_scenario", id="scenario_1", tags=["apple", "banana"])
+
+    scenario_1 = _ScenarioManager._create(scenario_1_cfg)
+    scenario_2 = _ScenarioManager._create(scenario_2_cfg)
+
+    scenario_1.cycle = cycle_1
+    scenario_2.cycle = cycle_2
+
+    _ScenarioManager._set_primary(scenario_1)
+    _ScenarioManager._set_primary(scenario_2)
+
+    primary_scenarios_sorted_by_name = [scenario_2, scenario_1]
+    assert primary_scenarios_sorted_by_name == _ScenarioManager._get_primary_scenarios(sorted=True, sort_key="name")
+
+    primary_scenarios_sorted_by_id = [scenario_2, scenario_1]
+    assert primary_scenarios_sorted_by_id == _ScenarioManager._get_primary_scenarios(sorted=True, sort_key="id")
+
+    primary_scenarios_sorted_by_creation_date = [scenario_1, scenario_2]
+    assert primary_scenarios_sorted_by_creation_date == _ScenarioManager._get_primary_scenarios(
+        sorted=True, sort_key="creation_date"
+    )
+
+    primary_scenarios_sorted_by_tag = [scenario_2, scenario_1]
+    assert primary_scenarios_sorted_by_tag == _ScenarioManager._get_primary_scenarios(sorted=True, sort_key="tags")
```

Comment:
Thank you for showing how configuring and creating scenarios is done in a more representative way. Indeed, it would be interesting to look at what happens when two scenarios have the same value. I will have a look at this!

Reply:
Hello @jrobinAV, 

Currently, I am busy implementing your recommended changes. However, I have some issues with using the representative way of configuring and creating scenarios, as now my primary scenarios seems to contain dates as well as scenarios when I print it. Hence, I think it would be nice if you can check whether I am doing it correctly. 
Here is what I currently have for the test function:
```suggestion
    def test_get_primary_scenarios_sorted():
    scenario_1_cfg = Config.configure_scenario(id="scenario_1", frequency=Frequency.DAILY)
    scenario_2_cfg = Config.configure_scenario(id="scenario_2", frequency=Frequency.DAILY)

    scenario_1 = _ScenarioManager._create(scenario_1_cfg, name="B_scenario")
    scenario_2 = _ScenarioManager._create(scenario_2_cfg, name="A_scenario")
    scenario_3 = _ScenarioManager._create(scenario_2_cfg, name="C_scenario")

    _ScenarioManager._set_primary(scenario_1)
    scenario_1.tags = ["banana", "kiwi"]
    _ScenarioManager._set_primary(scenario_2)
    scenario_2.tags = ["apple", "banana"]
    _ScenarioManager._set_primary(scenario_3)
    scenario_3.tags = ["banana", "kiwi"]

    print("all scenarios: ", _ScenarioManager._get_all)
    primary_scenarios = _ScenarioManager._get_primary_scenarios()
    print("primary scenarios: ", primary_scenarios)

    primary_scenarios_sorted_by_name = [scenario_2, scenario_1, scenario_3]
    print(_ScenarioManager._sort_scenarios(primary_scenarios, sort_key="name"))
    assert primary_scenarios_sorted_by_name == _ScenarioManager._sort_scenarios(primary_scenarios, sort_key="name")

    primary_scenarios_sorted_by_id_version1 = [scenario_1, scenario_2, scenario_3]
    primary_scenarios_sorted_by_id_version2 = [scenario_1, scenario_3, scenario_2]
    result_sort_primary_scenarios_by_id = _ScenarioManager._sort_scenarios(primary_scenarios, sort_key="id")
    print("result of sort primary scenarios by id: ")
    print(result_sort_primary_scenarios_by_id)
    assert (primary_scenarios_sorted_by_id_version1 == result_sort_primary_scenarios_by_id or 
            primary_scenarios_sorted_by_id_version2 == result_sort_primary_scenarios_by_id)

    primary_scenarios_sorted_by_creation_date = [scenario_1, scenario_3, scenario_2]
    assert primary_scenarios_sorted_by_creation_date == _ScenarioManager._sort_scenarios(primary_scenarios, 
                                                                                         sort_key="creation_date")

    primary_scenarios_sorted_by_tags_version1 = [scenario_2, scenario_1, scenario_3]
    primary_scenarios_sorted_by_tags_version2 = [scenario_2, scenario_3, scenario_1]
    result_sort_primary_scenarios_by_tags = _ScenarioManager._sort_scenarios(primary_scenarios, sort_key="tags")
    print("result of sort primary scenarios by tags: ")
    print(result_sort_primary_scenarios_by_tags)
    assert (primary_scenarios_sorted_by_tags_version1 == result_sort_primary_scenarios_by_tags or 
            primary_scenarios_sorted_by_tags_version2 == result_sort_primary_scenarios_by_tags)

    primary_scenarios_sorted_by_name_descending_order = [scenario_3, scenario_1, scenario_2]
    assert primary_scenarios_sorted_by_name_descending_order == _ScenarioManager._sort_scenarios(
        primary_scenarios, descending=True, sort_key="name"
    )
```

I figured out that my implementation of the sort mechanism is not predictable when handling two scenarios with the same value, for instance the same name. So, if scenario A and B have the same name, the returned sorted list might either be [A, B] or [B, A]. However, I think this is fine. What do you think. I appreciate your help!

Reply:
1. Two scenarios from the same cycle cannot be primary simultaneously. In your example, your 3 scenarios are from the same cycle. So when you promote `scenario_3` as primary, the two others are demoted as not primary anymore.

    To make them all 3 primary, they must belong to different cycles. You can pass different days as `creation_date` when creating the scenarios. Something like that:

    ``` python
    scenario_1 = _ScenarioManager._create(scenario_1_cfg, name="B_scenario")
    scenario_2 = _ScenarioManager._create(scenario_2_cfg, name="A_scenario", creation_date=datetime.datetime.now() + datetime.timedelta(days=2))
    scenario_3 = _ScenarioManager._create(scenario_2_cfg, name="C_scenario", creation_date=datetime.datetime.now() + datetime.timedelta(days=4))
    ```

2. About the unpredictability: We can use a second criterion in case of an equal value. The `id` is guaranteed unique, so we can use it as the second criterion to make it predictable. It is made of the config_id concatenated with a random unique id.
    
    Something like the following should do the job:

    ``` python
    if sorted:
        if sort_key in ["name", "creation_date", "tags"]:
            scenarios.sort(key=lambda x: (getattr(x, sort_key), x.id)
        elif sort_key == "id":
            scenarios.sort(key=lambda x: x.id
        else:
            scenarios.sort(key=lambda x: (x.name, x.id)
    ```
    In the test, we need to determine at runtime whether scenario_2.id or scenario_3.id is 'smaller'.
    
I Hope this can help. Please tell me if it does make sense to you.